### PR TITLE
feat(active) support map headers

### PR DESCRIPTION
**Pull request by mayocream**
## Summary

- change `active.headers` values to one or more lists of values indexed by header name.
- deprecate array headers.
- <del>escape characters in HTTP headers (See https://www.rfc-editor.org/rfc/rfc7230#page-9)</del>

### CODE REVIEW

File: lib/resty/healthcheck.lua
Comment (flrgh) [inline lib/resty/healthcheck.lua:975]: I don't recall off the top of my head, but it might be worth double-checking if the `typedefs.headers` input that we are expecting will possibly include tables in the header values. For instance, something like this:

```lua
{
  foo = "bar",
  multi = {"baz", "bat"},
}
```

...thus resulting in the following in the request text:

```
foo: bar
multi: baz
multi: bat
```

Actually that might even be the default "shape" of `typedefs.headers` (`map<string, string[]>`), now that I think of it.

Comment (mayocream) [reply lib/resty/healthcheck.lua]: Thanks! solved I think

File: lib/resty/healthcheck.lua
Comment (flrgh) [inline lib/resty/healthcheck.lua:963]: Are `self.checks.active.headers` ever mutated after the healthcheck is created? If not, maybe we should consider caching the final headers string somewhere in `self` now that we're doing a lot more work to create it.

File: lib/resty/healthcheck.lua
Comment (flrgh) [inline lib/resty/healthcheck.lua:977]: some trailing whitespace here:

```suggestion
              headers[idx] = escape_uri(k) .. ": " .. escape_uri(item)
```

File: lib/resty/healthcheck.lua
Comment (flrgh) [inline lib/resty/healthcheck.lua:42]: `ngx.escape_uri` might not be the proper function for this, as it will percent-encode and escape many values that are valid in an http header.

The more I think about this, maybe I was being hasty in recommending header sanitizing. Surprisingly (to me at least), a quick look around reveals that there are several examples of popular lua-resty-* code that takes header input but makes no attempt at sanitizing them:

* [lua-resty-http](https://github.com/ledgetech/lua-resty-http/blob/7552ac4a20d91240c03a6df111d94fc237c06d37/lib/resty/http.lua#L336-L350)
* [lua-resty-websocket](https://github.com/openresty/lua-resty-websocket/blob/34f6b41e538151e7a5f013ff9709c1708a9958fa/lib/resty/websocket/client.lua#L171-L175)

While I still _personally_ think we ought to strip `\r\n` from the header values, maybe someone will cite the above as a reason why we shouldn't.

So in the interest of not blowing up the scope on this PR, I think it might be better to punt on the sanitize task (remove the `ngx.escape_uri()` calls). I'd love to hear from someone else on the team if there's any more context to this topic though.

What do you think?

Comment (mayocream) [reply lib/resty/healthcheck.lua:42]: We can leave the validation works to the users, just like Kong uses a custom validator for HTTP header keys.
https://github.com/Kong/kong/blob/08b5349be535d712d5a65415f4c08f02742385ba/kong/db/schema/typedefs.lua#L545-L554
IMO, this library should only accept pre-formatted values.

**Review by flrgh — APPROVED** [review]
:+1: this looks good to me, though it might be wise to get a second review from somebody who has done more work in this repo (@locao or @Tieske maybe?)

**Comment by mayocream** [comment]
And a new release is needed for Kong I think?
- https://github.com/Kong/kong/pull/8255

**Pull request merged**
